crates/lace-bin/src/typechecking/ty.rs

```diff
@@ -84,42 +87,29 @@ impl<'a> PartialType<'a> {
     pub const Unit: Self = Self::Tuple(&[]);
 }
 
+impl<'a> From<TypeVariable> for PartialType<'a> {
+    fn from(value: TypeVariable) -> Self {
+        Self::Variable(value)
+    }
+}
+
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
 pub struct TypeVariable(pub usize);
 
 /// generates new type variables in increasing order.
 pub struct TypeVariableGenerator {
-    curr: usize,
+    curr: AtomicUsize,
```

Comment:
I wonder if ultimately the `Generator` should be per function too. I think we _should_ be able to resolve all type vars per function and not have any leak out right?

Reply:
not the global types. But yes, once we resolved the global tyvars each generator could be split in theory.